shared/sbl.js

```diff
@@ -0,0 +1,113 @@
+'use strict'
+
+type TypeFilter = (domain: string, selector: string, data: Array<*>) => ?boolean
+type TypeDomain = {
+  selectors: {[string]: Function},
+  domainFilters: Array<TypeFilter>,
+  selectorFilters: {[string]: TypeFilter},
+  data: Object
+}
+
+var domains: {[string]: TypeDomain} = {}
+var globalFilters: Array<TypeFilter> = []
+var pendingCalls = {} // allows sbl to be called regardless of module load order
+
+function sbl (path: string, ...data?: Array<*>) {
+  var [, name, selector] = /^([^/]+)(.+)/.exec(path)
+  var domain = domains[name]
+  if (!domain) {
+    // a call might happen to a selector for domains that haven't been registered
+    // yet, due to module load order. So we save them and run them later.
+    if (!pendingCalls[name]) pendingCalls[name] = []
+    pendingCalls[name].push([path, data])
+    return
+  }
+  // Filters can perform additional functions, and by returning `false` they
+  // can prevent the execution of a selector.
+  for (let gf of globalFilters) {
+    if (gf(name, selector, data) === false) return
+  }
+  for (let df of domain.domainFilters) {
+    if (df(name, selector, data) === false) return
+  }
+  var sf = domain.selectorFilters[selector]
+  if (sf && sf(name, selector, data) === false) return
+  return domain.selectors[selector].call(domain, ...data)
+}
+
+sbl.registerDomain = function (domain: string, selectors: {[string]: Function}) {
+  if (domains[domain]) throw new Error(`${domain} exists`)
+  domains[domain] = {selectors, data: {}, domainFilters: [], selectorFilters: {}}
+  // run any pendingCalls for this domain
+  if (pendingCalls[domain]) {
+    for (var [path, data] of pendingCalls[domain]) {
+      sbl(path, ...data)
+    }
+    delete pendingCalls[domain] // see sbl.throwIfPending
+  }
+}
+
+sbl.addDomainFilter = function (domain: string, filter: TypeFilter) {
+  domains[domain].domainFilters.push(filter)
+}
+
+sbl.setSelectorFilter = function (domain: string, selector: string, filter: TypeFilter) {
+  if (domains[domain].selectorFilters[selector]) {
+    throw new Error(`${domain}/${selector} filter exists`)
+  }
+  domains[domain].selectorFilters[selector] = filter
+}
+
+sbl.addGlobalFilter = function (filter: TypeFilter) {
+  globalFilters.push(filter)
+}
+
+// call at end of sbl setup, right before main run loop of app
+sbl.throwIfPending = function () {
```

Comment:
I don't understand this function. Why would there be pending calls if we have not started the application? Further, why would we need to call it in main rather than at the end of this object's definition if its being called before the application is started? Why would it be a problem if there were pending calls?